### PR TITLE
miniupnpd: add force_forwarding option support

### PR DESCRIPTION
**Pull request by ptpt52**
This option replace the 'ext_ip_reserved_ignore', and replace the
patch with new one

as addr_is_reserved() behavior should not be changed, so drop the
ext_ip_reserved_ignore

use a force_forwarding option to make the port forwarding force to
work when the router is behind NAT

Maintainer: me / @\<github-user> (find it by checking history of the package Makefile)
Compile tested: (put here arch, model, OpenWrt version)
Run tested: (put here arch, model, OpenWrt version, tests done)

Description:

### CODE REVIEW

**Comment by ptpt52** [comment]
need some tested

**Comment by ptpt52** [comment]
https://github.com/openwrt/packages/issues/15258
may related

**Review by BKPepe — CHANGES_REQUESTED** [review]
Even though I appreciate your work, I am not in favor to merge this because what happened here: https://github.com/openwrt/packages/issues/15258

We can not carry patches, which are refused to be part of upstream and are causing several issues, and then they are reported upstream, which they need to solve and refuse it as they are right about it.

If you want to have this merged, send this upstream, cooperate with them to be sure that it is accepted, and once it is merged, it can be backported here or you can wait until it's part of a new release.

For now, I am closing this to prevent the situation, which happened in the past. You should really remove the patch which you created. Because this is not an acceptable state. Users are confused about what to do with the error message, which is cryptic and does not say many details what to do.

**Comment by BKPepe** [comment]
Even though, I appreciate your work, I am not in favor to merge this because what happened here: https://github.com/openwrt/packages/issues/15258

We can not carry patches, which are refused to be part of upstream and they are causing several issues and then they are reported to upstream, which they needs to solve.

If you want to have this merged, send this upstream, cooperate with them to be sure that it is accepted and once it is merged, it can be backported here or you can wait until its part of a new release.

**Comment by Neustradamus** [comment]
@openips: What do you think about this PR from @ptpt52?
It replaces "301-ext_ip_reserved_ignore.patch"...

**Comment by openips** [comment]
@Neustradamus I test this pr in openwrt package and find not work， i can report bug here，thks a lot

**Comment by openips** [comment]
@ptpt52 大神 最新的补丁有问题  force_forwarding失效 具体描述请看 [16480](https://github.com/openwrt/packages/issues/16480) 有时间修复一下 需要提供测试请联系我 谢谢

**Comment by ptpt52** [comment]
@openips  try to enable stun
```
upnpd.config.use_stun='1'
upnpd.config.stun_host='iphone-stun.strato-iphone.de'
upnpd.config.stun_port='3478'
```

**Comment by ptpt52** [comment]
specify the wan interface:
```
upnpd.config.external_iface='wan or what'
```

**Comment by openips** [comment]
@ptpt52 我还是习惯您之前的301补丁的工作方式 非常省心 能否修改一下

**Comment by ptpt52** [comment]
try this one: https://github.com/x-wrt/packages/commit/62b923a284d899d02895af9006f5bd991246e89c
@openips

**Comment by openips** [comment]
好的 我编译一下试试看 结果稍后反馈

**Comment by openips** [comment]
image
image
@ptpt52 @Neustradamus x-wrt/packages@62b923a worked very well ，thks a lot

**Comment by Neustradamus** [comment]
@openips: Nice!

Thanks @ptpt52 :)
Can you create a new PR?

File: net/miniupnpd/files/upnpd.config
Comment (stangri) [inline net/miniupnpd/files/upnpd.config:5]: Please don't make this option default. Previously when you made `ext_ip_reserved_ignore` default, it resulted in cryptic syslog messages for miniupnpd users outside of China.

**Comment by Neustradamus** [comment]
@ptpt52: Important: Please do new PRs for 19.07.x/21.02.x/master branches.
- https://github.com/openwrt/packages/commits/master/net/miniupnpd
- https://github.com/openwrt/packages/commits/openwrt-21.02/net/miniupnpd
- https://github.com/openwrt/packages/commits/openwrt-19.07/net/miniupnpd

Thanks a lot in advance.

PS: @neheb: Too late to have same miniupnp version in 18.06.x?

**Comment by champtar** [comment]
Please do a PR for master only first, then we will backport once reviewed ;)

**Comment by Neustradamus** [comment]
@ptpt52: Important: Please do new PRs for 18.06.x/19.07.x/21.02.x/master branches.
- https://github.com/openwrt/packages/commits/master/net/miniupnpd
- https://github.com/openwrt/packages/commits/openwrt-21.02/net/miniupnpd
- https://github.com/openwrt/packages/commits/openwrt-19.07/net/miniupnpd
- https://github.com/openwrt/packages/commits/openwrt-18.06/net/miniupnpd

Thanks a lot in advance.

PS: @neheb: Too late to have same miniupnp version in 18.06.x/19.07.x/21.07.x/master?

**Comment by skill7899** [comment]
it not work at 23.03 fw4，focre_forwarding or ext_ip_reserved_ignore  Illegal parameter when i add in /var/etc/miniupnp.conf。 
invalid option in file /var/etc/miniupnpd.conf line 11 : ext_ip_reserved_ignore=no
invalid option in file /var/etc/miniupnpd.conf line 11 : focre_forwarding=yes
@ptpt52

**Comment by ptpt52** [comment]
> it not work at 23.03 fw4，focre_forwarding or ext_ip_reserved_ignore Illegal parameter when i add in /var/etc/miniupnp.conf。 invalid option in file /var/etc/miniupnpd.conf line 11 : ext_ip_reserved_ignore=no invalid option in file /var/etc/miniupnpd.conf line 11 : focre_forwarding=yes @ptpt52

these two config options never supported  by upstream miniupnpd

**Comment by skill7899** [comment]
> > it not work at 23.03 fw4，focre_forwarding or ext_ip_reserved_ignore Illegal parameter when i add in /var/etc/miniupnp.conf。 invalid option in file /var/etc/miniupnpd.conf line 11 : ext_ip_reserved_ignore=no invalid option in file /var/etc/miniupnpd.conf line 11 : focre_forwarding=yes @ptpt52
> 
> these two config options never supported by upstream miniupnpd

Does it mean that I need to merge patch and compile myself？

**Comment by skill7899** [comment]
@ptpt52 
compile fail with ntfatbles and firewall4
Patch failed!  Please fix ./patches/301-options-force_forwarding-support.patch!
make[3]: Leaving directory '/data/cpl/packages/net/miniupnpd'
time: package/feeds/packages/miniupnpd/nftables/compile#0.18#0.08#0.23
    ERROR: package/feeds/packages/miniupnpd failed to build (build variant: nftables).

**Comment by rmilecki** [comment]
Our attempts to upstream this feature have failed, see:
* https://github.com/miniupnp/miniupnp/pull/511
* https://github.com/miniupnp/miniupnp/pull/741

It seems however a similar option (`ignore_private_ip`) was added in the upstream by https://github.com/miniupnp/miniupnp/pull/813 and is present in the 2.3.9 release.

**Pull request closed without merging**
